### DISCUSSION TITLE
generic constraint new with parameter requirements

### DISCUSSION THREAD

**Discussion by LupusInferni315** [dotnet/csharplang#6449] [Language Ideas] [answered]
In essence the ability to constrain a generic type with a specified constructor signature, 
Examples
```CSHARP
public class Foo<TBar> where TBar : new(string, float)
```
```CSHARP
public class Example<TFoo, TBar> where TFoo : new(string, TBar)
```
```CSHARP
public class Foo<TBar> where TBar : new(float=) // specifies an optional parameter
```

The benefit of having this as an option would be to allow object creation of a type with parameters, as opposed to just parameterless constructors. Where the equals sign enforces an must have optional parameter, an optional parameter in a constructor can be provided even when not required (albeit the generic type wouldn't be aware if the parameter is optional without the equals sign).

**Comment by TahirAhmadov** [comment, marked as answer]
https://github.com/dotnet/csharplang/discussions/769

**Reply by LupusInferni315** [reply]
I missed that while I was looking to see if it had already been suggested, thank you for pointing it out.

**Comment by HaloFour** [comment]
> Where the equals sign enforces an must have optional parameter, an optional parameter in a constructor can be provided even when not required (albeit the generic type wouldn't be aware if the parameter is optional without the equals sign).

I don't think this would work unless the constraint also included the default value for the optional parameter.  The compiler needs that in order to embed the appropriate value at the callsite at compile time, otherwise it's not possible to emit a call to that method.

**Reply by LupusInferni315** [reply]
I never actually thought about how an optional parameter actually worked, I guess I always assumed that it was only really accessed by the constructor itself not treated like a constant and "swapped out" when compiling

**Comment by pinkfloydx33** [comment]
I'm not a fan of this. There's no telling what a single `int` argument means. One class may use it to enforce a minimum value, another as an array size and yet another constructor may treat it as seconds. You need to know what the parameter actually means, in a generic context that doesn't really make sense. You're better off constructing the object yourself.

If you own the types, you might be able to add a static abstract method to act as a factory of sorts. Similar to the `Create` methods in System.Numerics etc.

**Reply by TahirAhmadov** [reply]
You are the second person I heard who made this argument. The key missing aspect here is that, the constructor constraint is not _designed_ to be used to limit which type is used; you use interface and/or base type constraints for that. The purpose of parameterized `new(...)` constraint is to allow invoking a constructor natively, without using reflection.

WRT static factory methods, there are downsides. Chief among which is like you mentioned - you can't do that if you don't own the types. But even if you do, it's extra boilerplate. And the benefits are not really there - the same way a ctor can take `int` which can have different meanings, a factory method can take `int` and you have to make sure you pass in the right value. Yes, ctor constraint doesn't tell you what the parameter name is, but even with factory methods, the name doesn't enforce anything, you still have to decide what to pass into it, and the same thing applies to ctors - ultimately, you have to decide what is passed into it.
